### PR TITLE
[WIP] Implementation of FUGW and UCOOT

### PR DESCRIPTION
**Pull request by 6Ulm**
## Types of changes

This PR is dedicated to the implementation of 
1.Fused Unbalanced GW (or more correctly, its lower bound) 
2.  (Fused) Unbalanced COOT.

Since their structures, it is enough to write a common template, then write a wrapper for each divergence. 
More precisely, we create a method called `fused_unbalanced_cross_spaces_divergence` , in which
- `reg_type="independent"` corresponds to (Fused) UCOOT. This yields `unbalanced_co_optimal_transport` method.
- `reg_type="joint"` corresponds to FUGW. This yields `fused_unbalanced_gromov_wasserstein` method.

We also allow for unregularized approximation of FUGW and UCOOT, i.e. $\varepsilon = 0$, thanks to the Majorization-Minization `ot.unbalanced.mm_unbalanced` and `ot.unbalanced.lbfgsb_unbalanced` L-BFGS-B methods.

This implementation also allows for $2$ types of marginal penalization: Kullback-Leibler divergence and squared L2 norm.  We also allow the cost to be sub-differentiable w.r.t the input matrices and reference distributions. This is implemented in `unbalanced_co_optimal_transport2` and `fused_unbalanced_gromov_wasserstein` methods.

## PR checklist

- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: ot/gromov/_utils.py [hunk 1]
```diff
@@ -797,3 +797,347 @@ def update_barycenter_feature(
...
+        Bregman divergence, either "kl" (Kullback-Leibler divergence) or "l2" (half-squared L2 divergence)
+    nx : backend, optional
+        If let to its default value None, a backend test will be conducted.
+
+    Returns
+    ----------
+    Bregman divergence between two product measures.
+    """
+
+    if nx is None:
```
Comment (cedricvincentcuaz) [inline ot/gromov/_utils.py:923]: test

File: ot/gromov/_utils.py [hunk 2]
```diff
@@ -797,3 +797,347 @@ def update_barycenter_feature(
...
+
+        weighted_wx = wx * nx.sum(pi1 * wx) / l2_pi1
+        weighted_wy = wy * nx.sum(pi2 * wy) / l2_pi2
+        weighted_wxy = wxy * nx.sum(pi * wxy) / l2_pi if reg_type == "joint" else wxy
+        weighted_w = (weighted_wx, weighted_wy, weighted_wxy)
+
+        new_rho = (rho_x * l2_pi1, rho_y * l2_pi2)
+        new_eps = eps * l2_pi if reg_type == "joint" else eps
+
+    elif divergence == "kl":
```
Comment (cedricvincentcuaz) [inline ot/gromov/_utils.py:1054]: test

File: ot/solvers.py [hunk 3]
```diff
@@ -242,13 +256,20 @@ def solve(M, a=None, b=None, reg=None, reg_type="KL", unbalanced=None,
         arr.append(a)
     if b is not None:
         arr.append(b)
+    if c is not None:
```
Comment (cedricvincentcuaz) [inline ot/solvers.py:259]: remove None test -> `arr = [M, a, b, c]`

File: ot/utils.py [hunk 4]
```diff
@@ -1043,26 +1043,23 @@ def potentials(self):
         This pair of arrays has the same shape, numerical type
         and properties as the input weights "a" and "b".
         """
-        if self._potentials is not None:
-            return self._potentials
-        else:
-            raise NotImplementedError()
+        return self._potentials
```
Comment (cedricvincentcuaz) [inline ot/utils.py:1046]: What's behind this change?

File: ot/utils.py [hunk 5]
```diff
@@ -1043,26 +1043,23 @@ def potentials(self):
...
 
     @property
     def potential_a(self):
```
Comment (cedricvincentcuaz) [inline ot/utils.py:1049]: What's behind this change?

File: ot/utils.py [hunk 6]
```diff
@@ -1043,26 +1043,23 @@ def potentials(self):
...
+            return None
 
     @property
     def potential_b(self):
         """Second dual potential, associated to the "target" measure "b"."""
         if self._potentials is not None:
             return self._potentials[1]
         else:
-            raise NotImplementedError()
+            return None
```
Comment (cedricvincentcuaz) [inline ot/utils.py:1062]: same

**Review by cedricvincentcuaz — CHANGES_REQUESTED** [review]
Thanks a lot @6Ulm for this impressive work.
Follows some remarks and comments to conclude the PR.
You also forgot to update `ot.gromov.__init__.py`

File: ot/gromov/_unbalanced.py [hunk 7]
```diff
@@ -0,0 +1,1049 @@
...
+
+    # set gradients
+    ucoot = log_ucoot["ucoot_cost"]
+    ucoot = nx.set_gradients(ucoot,
+                             (X, Y, wx_samp, wx_feat, wy_samp, wy_feat),
+                             (gradX, gradY, grad_wx_samp, grad_wx_feat, grad_wy_samp, grad_wy_feat)
+                             )
+
+    if log:
+        return ucoot, log_ucoot
```
[the remark this answers is not in the document]
Comment (6Ulm) [reply ot/gromov/_unbalanced.py:734]: already tested in `test_ucoot.py`

File: ot/gromov/_utils.py [hunk 8]
```diff
@@ -797,3 +798,342 @@ def update_barycenter_feature(
...
 
     return sum(list_features) * inv_p[:, None]
+
+
+############################################################################
+# Methods related to fused unbalanced GW and unbalanced Co-Optimal Transport.
+############################################################################
+
+def div_to_product(pi, a, b, pi1=None, pi2=None, divergence="kl", mass=True, nx=None):
+    r"""Calculate the Bregman divergence between an arbitrary measure and a product measure.
```
Comment (cedricvincentcuaz) [inline ot/gromov/_utils.py:808]: shorter -> Fast computation of the Bregman divergence between an arbitrary measure and a product measure. Only support for Kullback-Leibler and half-squared L2 divergences.

File: ot/gromov/_utils.py [hunk 9]
```diff
@@ -797,3 +798,342 @@ def update_barycenter_feature(
...
+    This implementation induces cheaper cost than the direct calculation.
+    Only support for Kullback-Leibler and half-squared L2 divergences.
+
+    For half-squared L2 divergence:
+    .. math::
```
Comment (cedricvincentcuaz) [inline ot/gromov/_utils.py:813]: check spacing + that equation holds on a page. I think there are rendering issues.

File: ot/gromov/_utils.py [hunk 10]
```diff
@@ -797,3 +798,342 @@ def update_barycenter_feature(
...
+
+    elif divergence == "l2":
+        res = (nx.sum(pi**2) + nx.sum(a**2) * nx.sum(b**2)
+               - 2 * nx.dot(a, nx.dot(pi, b))) / 2
+
+    return res
+
+
+def div_between_product(mu, nu, alpha, beta, divergence, nx=None):
+    r"""Calculate the Bregmain divergence between two product measures.
```
Comment (cedricvincentcuaz) [inline ot/gromov/_utils.py:882]: same comments than above

File: ot/gromov/_unbalanced.py [hunk 11]
```diff
@@ -0,0 +1,1051 @@
...
+        **kwargs_solver):
+
+    r"""Compute the fused unbalanced cross-spaces divergence between two matrices equipped
+    with the distributions on rows and columns. We consider two cases of matrix:
+    - (Squared) similarity matrix in Gromov-Wasserstein setting,
+    whose rows and columns represent the samples.
+    - Arbitrary-size matrix in Co-Optimal Transport setting,
+    whose rows represent samples, and columns represent corresponding features/dimensions.
+
+    Return the sample and feature transport plans between
```
Comment (cedricvincentcuaz) [inline ot/gromov/_unbalanced.py:34]: merge this sentence with the first one above

File: ot/gromov/_unbalanced.py [hunk 12]
```diff
@@ -0,0 +1,1051 @@
...
+        reg_marginals=10, epsilon=0, divergence="kl",
+        unbalanced_solver="mm", alpha=0, M_samp=None, M_feat=None,
+        rescale_plan=True, init_pi=None, init_duals=None,
+        max_iter=100, tol=1e-7, max_iter_ot=500, tol_ot=1e-7,
+        log=False, verbose=False, **kwargs_solve):
+
+    r"""Compute the unbalanced Co-Optimal Transport between two Euclidean point clouds
+    (represented as matrices whose rows are samples and columns are the features/dimensions).
+
+    Return the sample and feature transport plans between
```
Comment (cedricvincentcuaz) [inline ot/gromov/_unbalanced.py:400]: merge this sentence with the first one above.

File: ot/gromov/_unbalanced.py [hunk 13]
```diff
@@ -0,0 +1,1051 @@
...
+        reg_marginals=10, epsilon=0, divergence="kl",
+        unbalanced_solver="sinkhorn", alpha=0, M_samp=None, M_feat=None,
+        rescale_plan=True, init_pi=None, init_duals=None,
+        max_iter=100, tol=1e-7, max_iter_ot=500, tol_ot=1e-7,
+        log=False, verbose=False, **kwargs_solve):
+
+    r"""Compute the unbalanced Co-Optimal Transport between two Euclidean point clouds
+    (represented as matrices whose rows are samples and columns are the features/dimensions).
+
+    Return the unbalanced Co-Optimal Transport cost between
```
Comment (cedricvincentcuaz) [inline ot/gromov/_unbalanced.py:546]: merge this sentence with the first one above.

File: ot/gromov/_unbalanced.py [hunk 14]
```diff
@@ -0,0 +1,1051 @@
...
+        Cx, Cy, wx=None, wy=None, reg_marginals=10, epsilon=0,
+        divergence="kl", unbalanced_solver="mm", alpha=0,
+        M=None, init_duals=None, init_pi=None, max_iter=100,
+        tol=1e-7, max_iter_ot=500, tol_ot=1e-7,
+        log=False, verbose=False, **kwargs_solve):
+
+    r"""Compute the lower bound of the fused unbalanced Gromov-Wasserstein (FUGW) between two similarity matrices.
+    In practice, this lower bound is used interchangeably with the true FUGW.
+
+    Return the transport plan between
```
Comment (cedricvincentcuaz) [inline ot/gromov/_unbalanced.py:750]: same

File: ot/gromov/_unbalanced.py [hunk 15]
```diff
@@ -0,0 +1,1051 @@
...
+        Cx, Cy, wx=None, wy=None, reg_marginals=10, epsilon=0,
+        divergence="kl", unbalanced_solver="mm", alpha=0,
+        M=None, init_duals=None, init_pi=None, max_iter=100,
+        tol=1e-7, max_iter_ot=500, tol_ot=1e-7,
+        log=False, verbose=False, **kwargs_solve):
+
+    r"""Compute the lower bound of the fused unbalanced Gromov-Wasserstein (FUGW) between two similarity matrices.
+    In practice, this lower bound is used interchangeably with the true FUGW.
+
+    Return the lower bound of the fused unbalanced Gromov-Wasserstein cost between
```
Comment (cedricvincentcuaz) [inline ot/gromov/_unbalanced.py:893]: same

**Review by cedricvincentcuaz — COMMENTED** [review]
Thank you for your last updates, you are almost there ! follows small comments on the documentation to correct then we'll be able to merge with the main branch :)

**Pull request merged**